backend/optimize.ml

```diff
@@ -0,0 +1,249 @@
+(* CR gtulba-lecu for xclerc: I am thinking of having this functionality in the
+   dll. It feels useful, what do you think? *)
+let rec prev_at_most steps cell =
+  (* Convention: must try to go back at least one element *)
+  assert (steps > 0);
+  match DLL.prev cell with
+  | Some prev_cell ->
+    if steps = 1 then prev_cell else prev_at_most (steps - 1) prev_cell
+  | None -> cell
+
+let rec get_cells' (cell : Cfg.basic Cfg.instruction DLL.cell option) size lst =
+  match cell with
+  | Some cell -> (
+    match size with
+    | 0 -> List.rev lst
+    | size -> get_cells' (DLL.next cell) (size - 1) (cell :: lst))
+  | None -> List.rev lst
+
+let get_cells cell size =
+  assert (size > 0);
+  get_cells' (DLL.next cell) (size - 1) [cell]
+
+(** Logical condition for simplifying the following case: 
+  {| 
+    mov x, y 
+    mov y, x 
+  |}
+
+   In this case, the second instruction should be removed *)
+
+let remove_useless_mov (cell : Cfg.basic Cfg.instruction DLL.cell) =
+  match get_cells cell 2 with
+  | [fst; snd] -> (
+    let fst_val = DLL.value fst in
+    let snd_val = DLL.value snd in
+    match fst_val.desc with
+    | Op (Move | Spill | Reload) -> (
+      let fst_src, fst_dst = fst_val.arg.(0), fst_val.res.(0) in
+      match snd_val.desc with
+      | Op (Move | Spill | Reload) ->
+        let snd_src, snd_dst = snd_val.arg.(0), snd_val.res.(0) in
+        if are_equal_regs fst_src snd_dst && are_equal_regs fst_dst snd_src
+        then (
+          if !Flambda_backend_flags.cfg_peephole_optimize_track
+          then update_csv "remove_useless_mov";
+          DLL.delete_curr snd;
+          Some (prev_at_most go_back_const fst))
+        else None
+      | _ -> None)
+    | _ -> None)
+  | _ -> None
+
+(** Logical condition for simplifying the following case: 
+  {| 
+    <op 1> const1, r 
+    <op 2> const2, r 
+  |}
+
+  to: 
+  {|
+    <op 1> (const1 <op 2> const2), r 
+  |}
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+(* CR-soon gtulba-lecu for gtulba-lecu: implement the rest of Intop_imm cases
+   (only Imulh left to optimize if possible)
+
+   it may be the case that after Iadd, Isub folding the result is 0, treat this
+   case. *)
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1 : Mach.integer_operation), (op2 : Mach.integer_operation) with
+  | Mach.Iand, Mach.Iand -> Some (Mach.Iand, imm1 land imm2)
+  | Ior, Ior -> Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor -> Some (Mach.Ixor, imm1 lxor imm2)
+  | Ilsl, Ilsl ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsl, imm1 + imm2)
+    else None
+  | Ilsr, Ilsr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsr, imm1 + imm2)
+    else None
+  | Iasr, Iasr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Iasr, imm1 + imm2)
+    else None
```

Comment:
It is probably OK, but we should
discuss what happens if `imm`*i*
is out of bounds but the sum is
within bounds.

Reply:
I think the user may have written:
```
asr x, -4 ;; undefined
asr x, 65 ;; undefined
```
and we are rewriting it to `asr x, 61`
which is defined.

So, I am not saying the rewrite is
wrong as one could argue that we
are free to pick any value in this
case. But I would at least leave
a comment we are potentially
changing the value (a value the
user should not rely upon...).